torchmdnet/neighbors/neighbors_cuda.cu

```diff
@@ -0,0 +1,264 @@
+    if (row * (row - 1) > 2 * index)
+        row--;
+    return row;
+}
+
+template <typename scalar_t>
+__global__ void forward_kernel_brute(uint32_t num_all_pairs, const Accessor<scalar_t, 2> positions,
+                                     const Accessor<int64_t, 1> batch, scalar_t cutoff_lower2,
+                                     scalar_t cutoff_upper2, bool loop, bool include_transpose,
+                                     Accessor<int32_t, 1> i_curr_pair,
+                                     Accessor<int32_t, 2> neighbors, Accessor<scalar_t, 2> deltas,
+                                     Accessor<scalar_t, 1> distances, bool use_periodic,
+                                     triclinic::Box<scalar_t> box) {
+    const uint32_t index = blockIdx.x * blockDim.x + threadIdx.x;
+    if (index >= num_all_pairs)
+        return;
+    const uint32_t row = get_row(index);
+    const uint32_t column = (index - row * (row - 1) / 2);
+    if (batch[row] == batch[column]) {
+        const scalar3<scalar_t> pos_i{positions[row][0], positions[row][1], positions[row][2]};
+        const scalar3<scalar_t> pos_j{positions[column][0], positions[column][1],
+                                      positions[column][2]};
+        const auto delta = triclinic::compute_distance(pos_i, pos_j, use_periodic, box);
+        const scalar_t distance2 = delta.x * delta.x + delta.y * delta.y + delta.z * delta.z;
+        if (distance2 < cutoff_upper2 && distance2 >= cutoff_lower2) {
+            const int32_t i_pair = atomicAdd(&i_curr_pair[0], include_transpose ? 2 : 1);
+            // We handle too many neighbors outside of the kernel
+            if (i_pair + include_transpose < neighbors.size(1)) {
+                const scalar_t r2 = sqrt_(distance2);
+                neighbors[0][i_pair] = row;
+                neighbors[1][i_pair] = column;
+                deltas[i_pair][0] = delta.x;
+                deltas[i_pair][1] = delta.y;
+                deltas[i_pair][2] = delta.z;
+                distances[i_pair] = r2;
+                if (include_transpose) {
+                    neighbors[0][i_pair + 1] = column;
+                    neighbors[1][i_pair + 1] = row;
+                    deltas[i_pair + 1][0] = -delta.x;
+                    deltas[i_pair + 1][1] = -delta.y;
+                    deltas[i_pair + 1][2] = -delta.z;
+                    distances[i_pair + 1] = r2;
+                }
+            }
+        }
+    }
+}
+
+template <typename scalar_t>
+__global__ void add_self_kernel(const int num_atoms, Accessor<scalar_t, 2> positions,
+                                Accessor<int32_t, 1> i_curr_pair, Accessor<int32_t, 2> neighbors,
+                                Accessor<scalar_t, 2> deltas, Accessor<scalar_t, 1> distances) {
+    const int32_t i_atom = blockIdx.x * blockDim.x + threadIdx.x;
+    if (i_atom >= num_atoms)
+        return;
+    const int32_t i_pair = atomicAdd(&i_curr_pair[0], 1);
```

Comment:
The number of atoms is known, so each thread can compute `i_pair`. `atomicAdd` could be called just once by one thread.

Reply:
I made it so each the first thread in each block calls atomicAdd for the rest of the threads in the block.